### DISCUSSION TITLE
How to identify a website as mastodon server instance

### DISCUSSION THREAD

**Discussion by the100rabh** [mastodon/mastodon#30547] [Q&A]
Hi,

I am planning to build a firefox extension which will take me to my mastodon instance to subscribe when I visit someone else's mastodon page. For that I guess I first need to identify if that page is a mastodon instance. What is the best way to do it, anyone have any suggestions. 

I can think of following ways to find it.

1. Check if the URL is in the users format. http[s]://[instance]/@username. Create a list of known instances(from https://instances.social/api/doc/#api-Instances-ListInstances offline) and use that to identify 
2. Check if the URL is in the users format. http[s]://[instance]/@username and the page has css path "html body.app-body.theme-default.no-reduce-motion.layout-single-column div#mastodon.notranslate.app-holder". Maybe there is a better check than this though.
3. Check if the URL is in the users format. http[s]://[instance]/@username and ping the server API(eg https://[instance]/api/v1/accounts/username) to see if it exists and confirm it to be a mastodon server. 

Any help wrt to identifying the page being a mastodon users page would be helpful. 

To me option 1 seems the best approach because it does not depend on doing anything additional than creating a list from https://instances.social/api/doc/#api-Instances-ListInstances

**Comment by akibasocial** [comment]
```regex
https://[a-zA-Z0-9.-]+/(@[a-zA-Z0-9_]+|web/accounts/[0-9]+)
```
This will match both patterns and from it you can get the `base address` and `username` or `Id`
- `https://[instance]/@[username]`
- `https://[instance]/web/accounts/[user_id]`

Then you can make a request to the lookup endpoint of the api and check if it's a valid profile
- [GET /api/v1/accounts/lookup HTTP/1.1](https://docs.joinmastodon.org/methods/accounts/#lookup)

**Reply by the100rabh** [reply]
I am wondering if I can avoid that extra API call in some manner. Maybe use the css selector ".server-banner__introduction > span:nth-child(1) > a:nth-child(2)" to see if it has a link to joinmastodon.org.  Would that work ?

**Comment by trwnh** [comment]
Are you trying to detect fediverse profiles, or specifically Mastodon profile pages? If you want to support non-Mastodon, then there won't be a consistent "users format URL".

It sounds like the real goal is to be able to remote follow. Is that correct? If so, then Mastodon supports a "remote interaction" flow at `https://mastodon.social/authorize_interaction?uri={uri}`. This is exposed via WebFinger and the link `rel`ation of `http://ostatus.org/schema/1.0/subscribe`. Your Firefox extension could simply plug the current URL into the template when you click the button; this would rely on the user knowing that they are looking at a fediverse profile currently.

**Comment by shaedrich** [comment]
You could call `/.well-known/nodeinfo` which will respond with something like this:
```json
{
  "links": [
    {
      "rel": "http://nodeinfo.diaspora.software/ns/schema/2.0",
      "href": "https://mastodon.online/nodeinfo/2.0"
    }
  ]
}
```

if you call `/nodeinfo/2.0` from `$.links[0].href`, you'll get in return something like
```json
{
  "version": "2.0",
  "software": {
    "name": "mastodon",
    "version": "4.3.0-nightly.2024-08-30"
  },
  "protocols": [
    "activitypub"
  ],
  "services": {
    "outbound": [],
    "inbound": []
  },
  "usage": {
    "users": {
      "total": 190561,
      "activeMonth": 13239,
      "activeHalfyear": 24093
    },
    "localPosts": 9599337
  },
  "openRegistrations": false,
  "metadata": {
    "nodeName": "Mastodon",
    "nodeDescription": "A newer server operated by the Mastodon gGmbH non-profit"
  }
}
```
where under `$.software`, you'll see, that it's mastodon.
